src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.cs

```diff
@@ -0,0 +1,145 @@
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.SyncNamespace
+{
+    internal abstract partial class AbstractSyncNamespaceService<TNamespaceDeclarationSyntax, TCompilationUnitSyntax> :
+       ISyncNamespaceService
+       where TNamespaceDeclarationSyntax : SyntaxNode
+       where TCompilationUnitSyntax : SyntaxNode
+    {
+        public async Task<ImmutableArray<CodeAction>> GetRefactoringsAsync(
+            Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            var state = await State.CreateAsync(this, document, textSpan, cancellationToken).ConfigureAwait(false);
+            if (state == null)
+            {
+                return default;
+            }
+
+            return CreateCodeActions(this, state);
+        }
+
+        public abstract bool TryGetReplacementReferenceSyntax(
+            SyntaxNode reference, ImmutableArray<string> newNamespaceParts, ISyntaxFactsService syntaxFacts, out SyntaxNode old, out SyntaxNode @new);
+
+        protected abstract string EscapeIdentifier(string identifier);
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="root"></param>
+        /// <param name="declaredNamespaceParts"></param>
+        /// <param name="targetNamespaceParts"></param>
+        /// <returns></returns>
+        protected abstract SyntaxNode ChangeNamespaceDeclaration(
+            SyntaxNode root, ImmutableArray<string> declaredNamespaceParts, ImmutableArray<string> targetNamespaceParts);
+
+        protected abstract IReadOnlyList<SyntaxNode> GetMemberDeclarationsInContainer(SyntaxNode compilationUnitOrNamespaceDecl);
+
+        /// <summary>
+        /// Determine if this refactoring should be triggered based on current cursor position and if there's any partial 
+        /// type declarations. It should only be triggered if the cursor is:
+        ///     (1) in the name of only namespace declaration
+        ///     (2) in the name of first declaration in global namespace if there's no namespace declaration in this document.
+        /// </summary>
+        /// <returns>
+        /// A tuple with two items:
+        ///     1. a boolean value indicates if the refactoring should be triggered.
+        ///     2. the only namespace declaration node in the document, null if none or more than one.
+        /// </returns>
+        protected abstract Task<(bool shouldTrigger, TNamespaceDeclarationSyntax singleNamespaceDecl)> ShouldPositionTriggerRefactoringAsync(
+            Document document, int position, CancellationToken cancellationToken);
+
+        protected SyntaxAnnotation WarningAnnotation
+            => CodeActions.WarningAnnotation.Create(
+                FeaturesResources.Warning_colon_changing_namespace_may_produce_invalid_code_and_change_code_meaning);
```

Comment:
1. make static readonly.  
2. you have the subclass add this annotation.  But it seems more correct to have this type responsible for adding it.

Reply:
1. Done.
2. I will leave it as is for now.